**NumPy**

- `Num -> Numerical Py -> Python`
- `NumPy -> Numerical Python => we are going to use to work with data which is in the form of dimension `

In [ ]:
# If we want to work with data which is in dimension or
# If we want to create data in the form of dimension we need to use NumPy

In [ ]:
import numpy as np


In [ ]:
len(dir(np))

532

**out of 532 Built_in_functions we are going to learn only the Built_in_functions which are used in AI (DA and DS)**





BUILT__IN__FUNCTION
- `array `
- `arange `
- `ndim `
- `shape `
- `size`
- `reshape `
- `reshape (-1,1) and (1,-1)`
- `zeros `
- `ones `
- `full `
- `eye `
- `diagonal `
- `ravel `
- `transpose `
- `concat`
- `random -> randint `
- `random -> randn`
- `random -> rand`
- `random -> random `
- `max `
- `sum `
- `min `
- `argmax `
- `argmin  `
- `repeat `
- `axis =1`
- `axis = 0 `
- `add and math function`
- `matmul `


In [ ]:
# array = which is used to create data in the form of dimension

f = np.array([1,9,6,5,43,8]) # 1d

print(f) # calling from Memory

print(f.ndim) # ndim (number of dimensions)

print(f.shape) # shape(rows and columns)

[ 1  9  6  5 43  8]
1
(6,)


In [ ]:
kite = np.array([1,9,6,5,43,8]) # 1D
print(kite)


[ 1  9  6  5 43  8]


In [ ]:
# 2d
import numpy as np

a = np.array([  [10,1,2]  , [3,6,9] , [2,4,6] ])

print(a)
print(a.ndim)
print(a.shape)
print(a.size) # (total values)

[[10  1  2]
 [ 3  6  9]
 [ 2  4  6]]
2
(3, 3)
9


In [ ]:
print(a[0][2])

2


In [ ]:
print(a[2][2])

6


In [ ]:
# changing 1D to 2D
a = np.array([10,11,14,17,19,16])


print(a)
print(a.ndim)
print(a.shape)

b = a.reshape(2,3)
print(b)
print(b.ndim)
print(b.shape)
print(b.size)

[10 11 14 17 19 16]
1
(6,)
[[10 11 14]
 [17 19 16]]
2
(2, 3)
6


In [ ]:
import numpy as np
z = np.array([1,2,3,4,5,6,7,8,9,10])

print(z)
print(z.size)
print(z.shape)
print(z.ndim)




[ 1  2  3  4  5  6  7  8  9 10]
10
(10,)
1


In [ ]:
# changing 1D to 2D

z = np.array([1,2,3,4,5,6,7,8])
print(z)
print(z.ndim)
print(z.shape)
print(z.size)




y = z.reshape(2,4)
print(y)
print(y.ndim)
print(y.size)
print(y.shape)

[1 2 3 4 5 6 7 8]
1
(8,)
8
[[1 2 3 4]
 [5 6 7 8]]
2
8
(2, 4)


In [ ]:
print(y[1][1])

6


In [ ]:
print(z[4])

5


In [ ]:
print(y[1][3])

8


In [ ]:
# arange = only 1D data with (n-1) logic

a = np.arange(19,31)
print(a)
print(a.ndim)

[19 20 21 22 23 24 25 26 27 28 29 30]
1


In [ ]:
# 1D to 2D

a = np.arange(1,71)
print(a)
print(a.ndim)
print(a.shape)

b = a.reshape(10,7)
print(b)
print(b.shape)
print(b.ndim)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70]
1
(70,)
[[ 1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14]
 [15 16 17 18 19 20 21]
 [22 23 24 25 26 27 28]
 [29 30 31 32 33 34 35]
 [36 37 38 39 40 41 42]
 [43 44 45 46 47 48 49]
 [50 51 52 53 54 55 56]
 [57 58 59 60 61 62 63]
 [64 65 66 67 68 69 70]]
(10, 7)
2


In [2]:
#zeros
import numpy as np

s = np.zeros((6,5))
print(s)


[[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]


In [ ]:
s.dtype

dtype('float64')

In [6]:
# ones

import numpy as np

s = np.ones((5,6))
print(s)




[[1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]]


In [15]:
#full

s = np.full((4, 5), 7)
print(s)

[[7 7 7 7 7]
 [7 7 7 7 7]
 [7 7 7 7 7]
 [7 7 7 7 7]]


In [8]:
#eye

import numpy as np

S = np.eye(3)

print(S)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [10]:
# diagonal function : its used to extract (or) create diagonal element in a matrix


s = np.diag([1,2,3,4,5,6])
print(s)

[[1 0 0 0 0 0]
 [0 2 0 0 0 0]
 [0 0 3 0 0 0]
 [0 0 0 4 0 0]
 [0 0 0 0 5 0]
 [0 0 0 0 0 6]]


In [15]:
# revalt : covert any dimension into 1D

s = np.array([[1,2,3],[4,5,6],[7,8,9]])
print(s)
print(s.shape)
print(s.ndim)

t = np.ravel(s)
print(t)
print(t.ndim)
print(t.shape)




[[1 2 3]
 [4 5 6]
 [7 8 9]]
(3, 3)
2
[1 2 3 4 5 6 7 8 9]
1
(9,)


In [18]:
# transpose : convert rows into column & column into rows

s = np.array ([[1,2,3],[4,5,6],[7,8,9]])
print(s)
print(s.ndim)
print(s.shape)
print(s.size)


t =np.transpose(s)
print(t)
print(t.ndim)
print(t.size)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
2
(3, 3)
9
[[1 4 7]
 [2 5 8]
 [3 6 9]]
2
9


In [26]:
# concat : merge 2 matrix
c = np.array([[12,13,14],[15,16,17],[18,19,20]])
f = np.array ([[34,4,5],[4,55,65],[6,7,45]])


print(c)
print(f)
print('========')


cf = np.concat([c,f])
print(cf)


[[12 13 14]
 [15 16 17]
 [18 19 20]]
[[34  4  5]
 [ 4 55 65]
 [ 6  7 45]]
[[12 13 14]
 [15 16 17]
 [18 19 20]
 [34  4  5]
 [ 4 55 65]
 [ 6  7 45]]


In [27]:
# random ->randint : get random values ho many you want you can write at last

a = np.random.randint(40,70)
print(a)

69


In [28]:
a = np.random.randint(40,70,6)
print(a)

[46 41 47 43 61 48]


In [29]:
# random -> rand -> random values between 0 to 1

a = np.random.rand(6,3)
print(a)

[[0.87284605 0.67296518 0.33374336]
 [0.05491479 0.64938941 0.70125161]
 [0.37797272 0.635145   0.37388054]
 [0.80510676 0.37167939 0.73266461]
 [0.62918578 0.57408983 0.21781264]
 [0.92716302 0.98330184 0.59327389]]


In [35]:


a = np.random.rand(6,5)

print(a)

[[0.56781629 0.52023748 0.05648594 0.87319002 0.00901184]
 [0.05196527 0.62076368 0.19898266 0.83044545 0.55549707]
 [0.63590185 0.20974903 0.56706009 0.68743148 0.08253984]
 [0.78680699 0.89485512 0.92076105 0.88687499 0.43721998]
 [0.75015161 0.2538067  0.55979065 0.48319346 0.27632758]
 [0.20735761 0.31288467 0.37720602 0.07449167 0.13804853]]


In [36]:
#random -> randn : random value between -1 to 1

s = np.random.randn(5,6)
print(s)

[[-0.02559932  1.09763174 -0.88747723  1.25771723  0.52966905 -0.28897088]
 [-0.68732208  1.51942368 -0.12714501 -0.31392888  0.29832471  0.36474596]
 [ 0.06330151 -2.51752549  0.46777189  1.73193528 -0.95578602  0.32224219]
 [ 0.23915752 -0.55273504 -0.69909793 -0.83831961 -1.21122055 -1.20358793]
 [ 0.12065214 -0.70612508  0.73559671 -0.94833883  1.29193212 -0.28025779]]


In [ ]:
# axis = 1 -> row
# axis = 0 -> column

In [38]:
b = np.array([ [10,30,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.sum(b))

77


In [39]:
b = np.array([ [10,30 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.sum(b , axis=0))

[20 43 14]


In [42]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.max(b , axis=0))

[10  7  8]


In [43]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.sum(b , axis=0))

[20 16 14]


In [44]:
b =np.array([[10,3,2],[1,6,4],[9,7,8]])
print(np.sum(b , axis =1))

[15 11 24]


In [45]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.max(b , axis=0))

[10  7  8]


In [51]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.max(b , axis=1))

[10  6  9]


In [54]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.min(b , axis=0))

[1 3 2]


In [53]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.min(b , axis=1))

[2 1 7]


In [55]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmax(b))

0


In [59]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmin(b))

3


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmax(b,axis=1))

[0 1 0]


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmin(b,axis=1))

[2 0 1]


In [ ]:
# repeat

a = np.arange(10,21)
b = a.reshape(1,-1)


c = np.repeat(b , 5 , axis=0)
print(c)

[[10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]]


In [ ]:
# some math operations

a = np.array([ [10,1,9],[ 3  ,6  ,7],[ 2 , 4 , 8] ])
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.add(a,b))

[[20  4 11]
 [ 4 12 11]
 [11 11 16]]


In [ ]:
# matmul

a = np.array([ [1,4] , [7,6] ])
b = np.array([ [4,3,2] , [1,8,6] ])

print(np.matmul(a,b))

[[ 8 35 26]
 [34 69 50]]
